### ISSUE TITLE
What is the output of Agent Neural Network? If there is a std, can we fix it manually? 

### ISSUE THREAD

**Issue by XueminLiu111** [tensorforce/tensorforce#785]
Hi there, I'm curious about the output of actor NN. 
In RL, the action is obtained by sampling from the output distribution of actor NN. Therefore, the output of actor NN must have something like mean and standard deviation if it is a Gaussian. We can also fix the std and let NN give us the mean. What is the setting in your library? Can we change it manually?

Besides, when we create the agent, we only need to provide the max and min value of actions. How do you choose the action if the sampled action outside the range? Do you select the boundary value or shrink the distribution?

Help appreciated!

**Comment by AlexKuhnle** [comment]
Oh, and you can change the default distribution settings if you explicitly specify the policy, as in the example at the top of the docs page:

```
Agent.create(
    ...
    policy=dict(distributions=dict(
        float=dict(type='gaussian', stddev_mode='global'),
        bounded_action=dict(type='beta')
    ))
    ...
)
```

You can either specify the distribution config per type (`float`) or per action name (`bounded_action`).

**Comment by XueminLiu111** [comment]
Thanks very much for your reply! I just want to make sure

>     * `stddev_mode` - `predicted` vs `global`: Whether to predict the stddev value from the input or whether to use a separate trainable variable (it's not possible right now to fix it to a constant value).

If it is  `predicted`, I check the docs which said ` linear transformation of the state embedding`, you mean 
(1) it adds one layer neural network to the agent NN? Is this the setting of `global`? 
(2) Or there is only one NN layer between the input(state) and the output(stddev)?

If it is  `global`, we can train the std like what we do on the mean?


> 
>     * `bounded_transform` - `tanh` vs `clipping`: If action space is bounded (recommended!), whether to use a tanh-transformed Gaussian or whether to use a standard Gaussian and clip the output values.
> 
For the bounded output, if we use `tanh` then we just add a tanh transform to the Gaussian samples and make it between (-1,1)? And then shift and multiple a magnitude to make it in the desired range? 
If it is `clipping`, we just clip the sample values if it outside the desired range?

**Comment by XueminLiu111** [comment]
And I also have a question on the batch_size when creating the agent.
In the doc, `batch_size (parameter, int > 0) – Number of episodes per update batch (required).`
If `batch_size=10`,  does this mean that we collect every 10 episodes and update the NN?
If we set `num_episodes=500`, then we'll update NN for 500/10=50 times?

**Comment by XueminLiu111** [comment]
Great, thanks so much for your help!

**Comment by AlexKuhnle** [comment]
No problem. Assume this can be closed then. :-)

**Comment by XueminLiu111** [comment]
>     * `predicted`: the main network processes a state into a "state embedding", and this setting adds an additional linear layer (plus softplus for positive values) to predict the `stddev` value per state. This is the same as what is done for the `mean`, so the network has "two heads", one for `mean = linear1(NN(state))`, one for `stddev = softplus(linear2(NN(state)))`.


Hi there, I still have a quick question about the `softplus` to predict the `stddev`. Do we need to do backpropogation on this `softplus` layer? Or we only need to do it on the `linear` layers?

Thanks!

**Comment by AlexKuhnle** [comment]
Yes, backpropagation is required, otherwise the agent wouldn't get any training signal as to how to optimise the stddev value. Why?

**Comment by AlexKuhnle** [comment]
To be precise: it's in particular the weights of `linear2` which need optimizing -- we could stop the gradients from backpropagating to the core network `NN`, if that is your question.

**Comment by XueminLiu111** [comment]
Thx. Because `softplus` is acting as an activation function to the output. I'm not sure do we need to do backpropagation on this `softplus` layer.

**Comment by AlexKuhnle** [comment]
We need to backpropagate through the `softplus` to get the gradients for the layers before, e.g. `linear2`. But `softplus` doesn't have any weights itself, so we don't do backprop specifically for the `softplus` layer. Does that clarify things?

**Comment by XueminLiu111** [comment]
Yes! Thanks very much!

**Comment by XueminLiu111** [comment]
> * `bounded_transform` - `tanh` vs `clipping`: If action space is bounded (recommended!), whether to use a tanh-transformed Gaussian or whether to use a standard Gaussian and clip the output values.


Hi there, I have another question about the `bounded_transform`.
When we bound actions, do we give these bounded actions both into the 
(1) env
(2) and the policy update -- for example, when we compute `log probability` of tanh actions to update policy NN.

Here is an example of using Tensorforce(0.5.2) to control flow pass cylinder with bounded actions
https://github.com/thw1021/Cylinder2DFlowControlGeneral/blob/ac5ea7fd304dedefa4489790e45c42ca9428440e/RobustDRLCylinder2DControl/Env2DCylinder.py#L1024

If we want to implement it using Pytorch for the policy NN as follows:
`class Policy(nn.Module):
    def __init__(self, state_dim, action_dim, hidden_size=(128, 128), activation='tanh', log_std=0):
        super().__init__()
        self.action_bound_cof = 8e-3
        self.is_disc_action = False
        if activation == 'tanh':
            self.activation = torch.tanh
        elif activation == 'relu':
            self.activation = torch.relu
        elif activation == 'sigmoid':
            self.activation = torch.sigmoid

        self.actor_hidden = nn.Sequential(
                nn.BatchNorm1d(236),
                init_layer(nn.Linear(236, 512)),
                nn.Tanh(),
                init_layer(nn.Linear(512, 256)),
                nn.Tanh(),
                init_layer(nn.Linear(256, 64)),
                nn.Tanh(),
            )
        self.action_mean = init_layer(nn.Linear(64, 4))
        self.action_log_std = nn.Parameter(torch.zeros(1, 4))

    def forward(self, x):
        x = self.actor_hidden(x)
        action_mean = self.action_mean(x)
        action_log_std = self.action_log_std.expand_as(action_mean)
        action_std = action_log_std.exp()
        return action_mean, action_log_std, action_std

    def select_action(self, x):
        action_mean, action_log_std, action_std = self.forward(x)
        action_std = action_std+ 1e-6 # added for NAN issue

        m = torch.distributions.Normal(action_mean, action_log_std.exp())

        action = m.sample()
        action2 = torch.tanh(action) * self.action_bound_cof
        # action2 = torch.clamp(action, min=-(8e-3), max=8e-3)
        action_mean_val = torch.mean(action2)
        action_bound = action2-action_mean_val

        return action_bound, action2`

 is it correct to use `torch.clamp` or `torch.tanh`?

**Comment by AlexKuhnle** [comment]
I can't say much about the PyTorch code, but in general: In Tensorforce, `bounded_transform` is applied to the actions as part of the distribution, so the bounded values are the ones that are passed to the environment as well as the ones used for updating. In case of `bounded_transform='tanh'`, the tanh-bijection is additionally taken into account for the update when computing the log-probability (since bounding technically modifies the distribution -- for tanh this can be taken into account analytically).

**Comment by XueminLiu111** [comment]
Got it, thanks very much!
